subsys/bluetooth/services/ias/ias_client.c

```diff
@@ -0,0 +1,147 @@
+#include <logging/log.h>
+#include <sys/byteorder.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/services/ias.h>
+
+#include "ias_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_IAS_CLIENT)
+#define LOG_MODULE_NAME bt_ias_client
+#include "../../common/log.h"
+
+#define IS_HANDLE_VALID(handle) ((handle) != 0x0000)
+
+static struct bt_ias_client_cb *ias_client_cb;
+static struct bt_ias_client client_list[CONFIG_BT_MAX_CONN];
+
+static struct bt_ias_client *client_by_conn(struct bt_conn *conn)
+{
+	return &client_list[bt_conn_index(conn)];
+}
+
+int bt_ias_alert_write(struct bt_conn *conn, enum bt_ias_alert_lvl alert_lvl)
+{
+	int err;
+	uint8_t lvl = (uint8_t)alert_lvl;
+
+	err = bt_gatt_write_without_response(conn,
+					client_by_conn(conn)->write.handle,
+					&lvl, sizeof(lvl), false);
+	if (err < 0) {
+		BT_ERR("IAS client alert level write failed");
+		return err;
+	}
+
+	return 0;
+}
+
+static uint8_t bt_ias_alert_lvl_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	CHECKIF(!attr) {
+		return BT_GATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_FOUND);
+	}
+
+	if (IS_HANDLE_VALID(attr->handle)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_HANDLE);
+	}
+
+	client_by_conn(conn)->write.handle = attr->handle;
+	return 0;
+}
+
+static uint8_t bt_ias_prim_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	int err = 0;
+	struct bt_gatt_service_val *data;
+	struct bt_ias_client *client = client_by_conn(conn);
+
+	if (!attr) {
+		BT_DBG("Discover complete\n");
+		return BT_GATT_ITER_STOP;
+	}
+
+	data = attr->user_data;
+
+	(void)memcpy(&client->uuid, BT_UUID_ALERT_LEVEL, sizeof(client->uuid));
+	client->discover.uuid = &client->uuid.uuid;
+	client->discover.start_handle = sys_cpu_to_le16(attr->handle + 1);
+	client->discover.end_handle = sys_cpu_to_le16(data->end_handle);
+	client->discover.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+	client->discover.func = bt_ias_alert_lvl_disc_cb;
+
+	err = bt_gatt_discover(conn, &client->discover);
+
+	if (err) {
+		BT_ERR("Characteristic discover failed\n");
+		ias_client_cb->discover(conn, err);
+	}
+	ias_client_cb->discover(conn, 0);
+
+	return BT_GATT_ITER_STOP;
+}
+
+int bt_ias_discover(struct bt_conn *conn)
+{
+	int err;
+	struct bt_ias_client *client = client_by_conn(conn);
+
+	(void)memcpy(&client->uuid, BT_UUID_IAS, sizeof(client->uuid));
+	client->discover.uuid = &client->uuid.uuid;
+	client->discover.func = bt_ias_prim_disc_cb;
+	client->discover.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+	client->discover.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	client->discover.type = BT_GATT_DISCOVER_PRIMARY;
+
+	err = bt_gatt_discover(conn, &client->discover);
+
+	if (err < 0) {
+		BT_ERR("Primary discover failed(err %d\n)", err);
+		ias_client_cb->discover(conn, err);
+	}
+
+	return err;
+}
+
+int bt_ias_client_cb_register(struct bt_ias_client_cb *cb)
```

Comment:
We should be able to store the callbacks as `const` 

```suggestion
int bt_ias_client_cb_register(const struct bt_ias_client_cb *cb)
```

Reply:
Done